### ISSUE TITLE
Deluxe Paint IV - Black stripe in Overscan mode

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#656]
In Overscan mode, DP4 shows a black vertical bar at the right which shouldn't be there:
Deluxe Paint IV v4.0 (1991)(Eletronics Arts)[a]

<img width="958" alt="Bildschirmfoto 2022-02-27 um 19 49 29" src="https://user-images.githubusercontent.com/12561945/155895674-3e27ce1c-4f0c-4887-b05c-03185828783b.png">

**Comment by dirkwhoffmann** [comment]
I think I've tracked down the problem. The first chunk of the toolbar (left of the gap) is drawn by the following BPL event: 

<img width="1016" alt="Bildschirmfoto 2022-03-20 um 15 41 12" src="https://user-images.githubusercontent.com/12561945/159168343-04d26c41-3606-4abc-8a65-00ea980f20aa.png">

([OE] means the shift registers for both the odd and even planes are bumped out.

The second chunk (right of the gap) is drawn by this event:

<img width="1016" alt="Bildschirmfoto 2022-03-20 um 15 42 00" src="https://user-images.githubusercontent.com/12561945/159168367-90ad9649-f762-4598-a181-d6cf6fe072b4.png">

On the real machine, I expect the horizontal counter to wrap over earlier (similar to what we see in the Copper) and DPaint seems to exploit this. 

Unfortunately, taking an early wrap-over into account isn't enough. To fix it in a clean way, a more fundamental change is needed. Right now, the drawing routine (PixelEngine) is invoked in the Agnus hsync handler which is technically wrong because the first part of a DMA line is still visible in the previous line. vAmiga should run the drawing routine somewhat later, at the beginning of the HSYNC area. If I remember correctly, UAE has undergone a similar change in the past. Because this is a too far-reaching change for v2.0, I'll probably make it a v2.1 featue.

**Comment by dirkwhoffmann** [comment]
Here is some experimental code to correctly align early drawing events (around the time when hpos wraps over): 

```
Sequencer::computeBplEvents(isize strt, isize stop, DDFState &state)
        ...
        isize jj = j > 16 ? j : HPOS_CNT + j;
        if ((jj & mask) == (agnus.scrollOdd & mask))  id = (EventID)(id | 1);
        if ((jj & mask) == (agnus.scrollEven & mask)) id = (EventID)(id | 2);
```

With this change applied, DPaint looks like this:

<img width="1050" alt="Bildschirmfoto 2022-05-20 um 13 12 34" src="https://user-images.githubusercontent.com/12561945/169516669-dec14b4c-a001-46d7-8526-6b6274991ea6.png">

The black vertical stripe is caused by another bug. It is likely related to #486.

**Comment by dirkwhoffmann** [comment]
Now sprites are visible inside the overscan area, too 😎:

<img width="1050" alt="Bildschirmfoto 2022-05-21 um 22 26 50" src="https://user-images.githubusercontent.com/12561945/169668022-4af4bfff-d231-4759-a0ec-46e480ed7f24.png">

Unfortunately, my fix is pretty much a hack, because at the time the sprite is drawn, the underlying playfield pixels haven't been synthesised yet. I've solved this by magically shuffling around various pixels at the end of each scanline. Originally, I wanted to solve the issue by drawing a scanline not at the end of a DMA line, but at the beginning of the HBLANK area. However, this causes lots of problems which finally made me jeopardise this idea.

Overall, I am unsure about the importance of supporting extreme overscan. Under normal circumstances, the part of the screen we're talking about is not visible. E.g., I cannot run DPaint on my real A500 in max overscan mode, because in this mode only a very few pixels of the toolbar is visible on my connected TFT. Of course, a real Amiga monitor might be completely different. I have no idea how my Scart to HDMI adapter determines the screen geometry.

Does anybody know demos, games or tools other than DPaint that utilise such an extreme overscan mode?

**Comment by mras0** [comment]
[Seven seas by Andromeda](https://www.pouet.net/prod.php?which=5373) mentioned in [this EAB thread](https://eab.abime.net/showthread.php?p=1080244#post1080244) (info from Toni in post 10) is probably a good bet. Also some more discussion [here](https://eab.abime.net/showthread.php?t=108314). There's also [this](http://eab.abime.net/showthread.php?t=45691) thread about maximum overscan.

I had to implement a bunch of hacks to get DPaintIV working, but I'm not sure they're actually appropriate for OCS (I have issues with 7 seas though, but haven't bothered fixing them).

On my A1200 (with a multisync monitor) in DPaintV I can see at least 368 lores pixels are there, but since I can't "squish" the display only move left or right it's not really usable.

**Comment by mras0** [comment]
I've tried to make a small overscan testcase:
[overscan2zip.zip](https://github.com/dirkwhoffmann/vAmiga/files/8749202/overscan2zip.zip)

On my A1200 the extreme left looks like this (Sorry for the poor quality and dirty monitor 😳):
![image](https://user-images.githubusercontent.com/9952017/169690464-9bf07c80-4c31-4dc2-981a-4631c0851942.png)
Right side:
![image](https://user-images.githubusercontent.com/9952017/169690477-6478f79f-c36a-48c6-af62-e420a12ac141.png)


In WinUAE 4.9.1 (OCS, extreme overscan) right side seems to match:
![image](https://user-images.githubusercontent.com/9952017/169690606-7ffa7c5e-3a5d-4837-b72c-f2f646efe094.png)

**Comment by mithrendal** [comment]
🤤 I like this !!!

**Comment by dirkwhoffmann** [comment]
Very useful test case! It exhibits very well where vAmiga is wrong:

<img width="1025" alt="Bildschirmfoto 2022-05-22 um 17 18 38" src="https://user-images.githubusercontent.com/12561945/169702654-7b6c4536-dbf8-4cc1-82cf-b62736695669.png">

My ViewSonic is not well suited for debugging overscan issues. It crops the image very early:

![IMG_2793](https://user-images.githubusercontent.com/12561945/169702954-b473abf5-b70d-40ee-9655-15b7d739d2d6.JPG)

**Comment by dirkwhoffmann** [comment]
Fixed in v2.1b2
